File: packages/core_dart/lib/src/format_scalar.dart
/// `formatScalar` — espejo 1:1 de `format-scalar.ts`.
///
/// Formato visible de un valor escalar según su behavior. Función pura:
/// `(value, def?) → String`. No produce widgets — solo el texto final. El
/// caller decide cómo lo envuelve. Determinismo cross-language: mismos
/// `(value, behavior)` → mismo output que el TS (ver format_scalar_test.dart).
library;

import 'dart:convert';

import 'field_def.dart';

const Map<String, String> _currencySymbols = {
  'EUR': '€', 'USD': '\$', 'GBP': '£', 'JPY': '¥',
};

bool _isEmpty(dynamic v) {
  if (v == null) return true;
  if (v is String) return v.trim().isEmpty;
  if (v is List) return v.isEmpty;
  return false;
}

/// String de un número como lo daría `String(n)` en JS (sin `.0` en enteros).
String _numStr(num v) {
  if (v is int) return v.toString();
  if (v == v.truncateToDouble()) return v.toInt().toString();
  return v.toString();
}

/// Inserta separador de miles '.' (es-ES) en una cadena de dígitos.
String _groupThousands(String digits) {
  final buf = StringBuffer();
  for (var i = 0; i < digits.length; i++) {
    if (i > 0 && (digits.length - i) % 3 == 0) buf.write('.');
    buf.write(digits[i]);
  }
  return buf.toString();
}

/// Formato es-ES con 2 decimales: 1234.5 → "1.234,50".
String _formatEs2dec(num value) {
  final neg = value < 0;
  final cents = (value.abs() * 100).round();
  final intPart = cents ~/ 100;
  final decPart = cents % 100;
  final s = '${_groupThousands(intPart.toString())},${decPart.toString().padLeft(2, '0')}';
  return neg ? '-$s' : s;
}

/// Formato visible para un valor escalar según su behavior.
String formatScalar(dynamic value, [FieldDefLike? def]) {
  if (_isEmpty(value)) return '';
  final b = def?.behavior;

  // Año entero: 2026 (sin separadores).
  if (b == 'year' && value is num) return _numStr(value);

  // Fecha ISO: 2026-05-24 → "24/5/2026" (es). Inválida → string original.
  if (b == 'iso_date' && value is String) {
    try {
      final d = DateTime.parse(value);
      return '${d.day}/${d.month}/${d.year}';
    } catch (_) {
      return value;
    }
  }

  // Currency: 19.99 → "19,99 €".
  if (b == 'currency' && value is num) {
    return '${_formatEs2dec(value)} ${_currencySymbols['EUR']}';
  }

  // Percentage: 75 → "75 %".
  if (b == 'percentage' && value is num) return '${_numStr(value)} %';

  // Rating: 4 → "★★★★☆".
  if (b == 'rating' && value is num) {
    const max = 5;
    final v = value.round().clamp(0, max).toInt();
    return '★' * v + '☆' * (max - v);
  }

  // Measurement: la unidad la define behaviorConfig del editor.
  if (b == 'measurement') {
    return value is num ? _numStr(value) : value.toString();
  }

  // Incremental (KRO-84 V2): pad + prefijo/sufijo opcionales (solo presentación).
  if (b == 'incremental' && value is num) {
    final cfg = def?.behaviorConfig ?? const <String, dynamic>{};
    final padRaw = cfg['pad'];
    final pad = (padRaw is num && padRaw > 0) ? padRaw.truncate() : 0;
    final prefix = cfg['prefix'] is String ? cfg['prefix'] as String : '';
    final suffix = cfg['suffix'] is String ? cfg['suffix'] as String : '';
    final n = value.truncate();
    final body = pad > 0
        ? n.abs().toString().padLeft(pad, '0')
        : n.abs().toString();
    return '${n < 0 ? '-' : ''}$prefix$body$suffix';
  }

  // ordinal_enum / enum: el valor ya es string del catálogo.
  if (value is String) return value;
  if (value is num) return _numStr(value);
  if (value is bool) return value ? 'sí' : 'no';

  // Fallback: stringify defensivo.
  try {
    return jsonEncode(value);
  } catch (_) {
    return value.toString();
  }
}

File: packages/core/src/format-scalar.ts
/**
 * `format-scalar.ts` — formateo visible de valores escalares según behavior.
 *
 * Función pura: dado un value + (opcional) FieldDefLike → string listo para
 * renderizar. No produce JSX/Widgets — solo el texto final. El caller decide
 * cómo lo envuelve (Studio en `<span>`, Flutter en `Text(...)`).
 *
 * Cubre los behaviors estables al 2026-05-27:
 *  - year, iso_date, currency, percentage, rating, measurement
 *  - fallback genérico para string/number/boolean
 *
 * **Determinismo cross-language**: para los mismos `(value, behavior)`,
 * el output debe ser idéntico en TS y en futuras impl. (Dart). Los tests
 * en `tests/format-scalar.test.ts` son la spec.
 *
 * Historia: extraído de `kromia-studio/src/components/album/recipes/
 * recipe-utils.tsx:461-506` en KRO-73 (migración B+).
 */

import type { FieldDefLike } from './types';

// ── Helpers internos ────────────────────────────────────────────────────

const CURRENCY_SYMBOLS: Record<string, string> = {
  EUR: '€', USD: '$', GBP: '£', JPY: '¥', CHF: 'CHF', CAD: '$', AUD: '$',
  MXN: '$', BRL: 'R$', ARS: '$', CNY: '¥', KRW: '₩', INR: '₹',
};

/** True si el valor es null/undefined/'' o array vacío. Privado al módulo. */
function isEmpty(v: unknown): boolean {
  if (v == null) return true;
  if (typeof v === 'string') return v.trim() === '';
  if (Array.isArray(v))      return v.length === 0;
  return false;
}

// ── API pública ─────────────────────────────────────────────────────────

/**
 * Formato visible para un valor escalar según su behavior.
 *
 * El símbolo de currency real lo gestiona la celda del editor con
 * behaviorConfig — el SDK solo conoce € por defecto. Las recetas
 * delegan la decisión a Studio (que sí sabe el config del field).
 *
 * @param value  El valor raw (número, string, boolean, etc.).
 * @param def    FieldDefLike con `behavior` opcional. Si no se provee,
 *               se aplica fallback genérico.
 * @returns      String listo para renderizar. `''` si el value está vacío.
 */
export function formatScalar(
  value: unknown,
  def?:  FieldDefLike,
): string {
  if (isEmpty(value)) return '';
  const b = def?.behavior;

  // Año entero: 2026 (sin separadores de miles).
  if (b === 'year' && typeof value === 'number') return String(value);

  // Fecha ISO: 2026-05-24 → "24/05/2026" en es-ES.
  if (b === 'iso_date' && typeof value === 'string') {
    try { return new Date(value).toLocaleDateString('es'); }
    catch { return value; }
  }

  // Currency: 19.99 → "19,99 €". KRO-198 — el código de moneda viene de
  // behaviorConfig.currency (ISO: EUR/USD/GBP/JPY/…); fallback EUR. Símbolo
  // SIEMPRE tras el número (formato es-ES uniforme, determinista cross-lang);
  // si el código no tiene símbolo conocido, se usa el propio código. JPY/KRW
  // sin decimales.
  if (b === 'currency' && typeof value === 'number') {
    const cfg  = (def?.behaviorConfig ?? {}) as Record<string, unknown>;
    const code = typeof cfg.currency === 'string' && cfg.currency.trim()
      ? cfg.currency.trim().toUpperCase() : 'EUR';
    const symbol  = CURRENCY_SYMBOLS[code] ?? code;
    const noCents = code === 'JPY' || code === 'KRW';
    const digits  = noCents ? 0 : 2;
    return `${value.toLocaleString('es-ES', { minimumFractionDigits: digits, maximumFractionDigits: digits })} ${symbol}`;
  }

  // Percentage: 75 → "75 %".
  if (b === 'percentage' && typeof value === 'number') return `${value} %`;

  // Rating: 4 → "★★★★☆".
  if (b === 'rating' && typeof value === 'number') {
    const max = 5;
    const v = Math.max(0, Math.min(max, Math.round(value)));
    return '★'.repeat(v) + '☆'.repeat(max - v);
  }

  // Measurement: 12.5 → "12.5 cm". KRO-198 — la unidad viene de
  // behaviorConfig.unit (string libre: cm/kg/ml/…); sin unidad → número plano.
  if (b === 'measurement') {
    const cfg  = (def?.behaviorConfig ?? {}) as Record<string, unknown>;
    const unit = typeof cfg.unit === 'string' && cfg.unit.trim() ? ` ${cfg.unit.trim()}` : '';
    return `${value}${unit}`;
  }

  // Incremental (contador / dorsal) — KRO-84 V2: zero-padding + prefijo/sufijo
  // OPCIONALES vía behaviorConfig. Solo PRESENTACIÓN: en BD se guarda el number
  // puro y el sort/búsqueda siguen sobre el número real. Sin config → número
  // plano (compat V1). Ej.: 7 + {pad:3, prefix:'HC-'} → "HC-007".
  if (b === 'incremental' && typeof value === 'number') {
    const cfg    = (def?.behaviorConfig ?? {}) as Record<string, unknown>;
    const pad    = typeof cfg.pad === 'number' && cfg.pad > 0 ? Math.trunc(cfg.pad) : 0;
    const prefix = typeof cfg.prefix === 'string' ? cfg.prefix : '';
    const suffix = typeof cfg.suffix === 'string' ? cfg.suffix : '';
    const n      = Math.trunc(value);
    const body   = pad > 0 ? String(Math.abs(n)).padStart(pad, '0') : String(Math.abs(n));
    return `${n < 0 ? '-' : ''}${prefix}${body}${suffix}`;
  }

  // ordinal_enum / enum: el valor ya es un string del catálogo.
  if (typeof value === 'string')  return value;
  if (typeof value === 'number')  return String(value);
  if (typeof value === 'boolean') return value ? 'sí' : 'no';

  // Fallback: stringify defensivo.
  try { return JSON.stringify(value); }
  catch { return String(value); }
}
